src/mesh/stats-mesh-adapter-decorator.ts

```diff
@@ -0,0 +1,50 @@
+import { MeshAdapter } from "./mesh-adapter.ts";
+import { MeshPlatform, MeshTenant } from "./mesh-tenant.model.ts";
+import { QueryStatistics } from "./query-statistics.ts";
+
+/**
+ * This adapter will try to fetch tenant data first from the local cache before
+ * it hits the real cloud platforms.
+ */
+export class StatsMeshAdapterDecorator implements MeshAdapter {
+  constructor(
+    private readonly meshAdapter: MeshAdapter,
+    private readonly platform: MeshPlatform,
+  ) {
+  }
+  async attachTenantCosts(
+    tenants: MeshTenant[],
+    startDate: Date,
+    endDate: Date,
+    stats: QueryStatistics,
+  ): Promise<void> {
+    return await stats.recordQuery(
+      this.platform,
+      async () =>
+        await this.meshAdapter.attachTenantCosts(
+          tenants,
+          startDate,
+          endDate,
+          stats,
```

Comment:
d: This is totally weird. I wanted to remark you that all the other meshAdapter implementations don't really require this parameter here, yet now the call signature of the methods is polluted with it. However as it turns out, e.g. the `AzureMeshAdapter` does not even implement this parameter (!) even though its required via the interface:

```
export interface MeshAdapter {
// ...
  attachTenantCosts(
    tenants: MeshTenant[],
    startDate: Date,
    endDate: Date,
    stats: QueryStatistics,
  ): Promise<void>;
```
but here, how does this even compile?

```
export class AzureMeshAdapter implements MeshAdapter {
  constructor(
    private readonly azureCli: AzureCliFacade,
    private readonly timeWindowCalculator: TimeWindowCalculator,
  ) {}

  async attachTenantCosts(
    tenants: MeshTenant[],
    startDate: Date,
    endDate: Date,
  ): Promise<void> {
// ...
```

but with me suggested change this wont be required anymore.